### PR TITLE
[DRAFT] Introduce PTQWrapper

### PR DESCRIPTION
**Pull request by mhs4670go**
This draft introduces PTQWrapper.

Related: #89
TICO-DCO-1.0-Signed-off-by: seongwoo <mhs4670go@naver.com>

### CODE REVIEW

File: tico/experimental/quantization/algorithm/ptq/examples/quantize_lienar.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# Copyright (c) 2025 Samsung Electronics Co., Ltd. All Rights Reserved
```
Comment (dayo09) [inline tico/experimental/quantization/algorithm/ptq/examples/quantize_lienar.py:1]: Typo in the filename: quantize_lienar -> quantize_linear

File: tico/experimental/quantization/algorithm/ptq/quant_config.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    """
+
+    default_dtype: DType = DType.uint(8)
+    default_observer: Type[ObserverBase] = MinMaxObserver
+    default_qscheme: QScheme = QScheme.PER_TENSOR_AFFINE
+    overrides: Mapping[str, Mapping[str, Any]] = field(default_factory=dict)
+
+    def get_kwargs(self, obs_name: str) -> Dict[str, Any]:
+        """
+        Return user-specified kwargs for *obs_name* inside **this** wrapper.
```
Comment (dayo09) [inline tico/experimental/quantization/algorithm/ptq/quant_config.py:80]: Q. do you use specific editor to parse this markdown-style enhancements? Actually it's hard to read in my plain vscode editor...

File: tico/experimental/quantization/ptq/examples/quantize_with_gptq.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+# -------------------------------------------------------------------------
+print("Applying GPTQ …")
+dataset = load_dataset("wikiText", "wikitext-2-raw-v1", split="test")
+q_m = prepare(model, GPTQConfig(), inplace=True)
+
+for d in tqdm.tqdm(dataset, desc="GPTQ calibration"):
+    ids = tokenizer(d["text"], return_tensors="pt").input_ids.to(DEVICE)
+    q_m(ids)  # observers gather weight stats
+
+q_m = convert(q_m, inplace=True)  # materialize INT-weight tensors
```
Comment (dayo09) [inline tico/experimental/quantization/ptq/examples/quantize_with_gptq.py:110]: It's great use of tqdm above but when running this script, `q_m = convert(q_m, inplace=True)` takes more time.
Could you add tqdm to quantizer.py too?

```py
        quantizers: Dict[str, Any] = {}
        for l_idx, layer in enumerate(tqdm.tqdm(target_layers)): # <-- to track the conversion in terminal.
```

File: tico/experimental/quantization/algorithm/ptq/examples/compare_ppl.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+# ── local quantization helpers ────────────────────────────────────────────────
+from tico.experimental.quantization.algorithm.ptq.quant_config import QuantConfig
+from tico.experimental.quantization.algorithm.ptq.utils import perplexity
+from tico.experimental.quantization.algorithm.ptq.wrappers.ptq_wrapper import PTQWrapper
+
+# -------------------------------------------------------------------------
+# 0. Global configuration
+# -------------------------------------------------------------------------
+MODEL_NAME = "meta-llama/Meta-Llama-3-1B"
```
[the remark this answers is not in the document]
Comment (mhs4670go) [reply tico/experimental/quantization/algorithm/ptq/examples/compare_ppl.py]: I just wrote it down for the example purpose because it is gated model. You should replace that string with valid ones.

File: tico/experimental/quantization/ptq/examples/quantize_with_gptq.py (hunk 3 above)
Comment (mhs4670go) [reply tico/experimental/quantization/ptq/examples/quantize_with_gptq.py:110]: Yeah, I think so. I kept thinking about it and ketp putting it off, lol. Thanks for the reivew:)

File: tico/experimental/quantization/algorithm/ptq/quant_config.py (hunk 2 above)
Comment (mhs4670go) [reply tico/experimental/quantization/algorithm/ptq/quant_config.py]: But, let's have some more discussion about preferences in the near future.

File: tico/experimental/quantization/ptq/wrappers/quant_elementwise.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+class QuantElementwise(QuantModuleBase):
+    """
+    Generic wrapper for *any* 1-to-1 element-wise op `y = f(x)`.
+
+    Sub-classes only need to implement:
+        • `FUNC`: a Callable that maps tensor→tensor
+    """
+
+    # subclass must set this
+    FUNC: Callable[[torch.Tensor], torch.Tensor] = staticmethod(lambda x: x)
```
Comment (dayo09) [inline tico/experimental/quantization/ptq/wrappers/quant_elementwise.py:36]: If it's must be set, why not "throw Exception"?

**Comment by mhs4670go** [comment]
> In case of masking, some channel io could all be set as -inf. In the case, will this ema collect logic work properly?

@dayo09 Good catch. Therefore, the mask is created with -120 value.

File: tico/experimental/quantization/ptq/wrappers/quant_elementwise.py (hunk 5 above)
Comment (mhs4670go) [reply tico/experimental/quantization/ptq/wrappers/quant_elementwise.py:36]: I've added `__init_subclass__` for this. Thanks.

File: tico/experimental/quantization/ptq/wrappers/quant_module_base.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            obs.collect(x.detach())
+            return x
+        if self._mode is Mode.QUANT:
+            return obs.fake_quant(x)
+        return x  # NO_QUANT
+
+    @abstractmethod
+    def _all_observers(self) -> Iterable[ObserverBase]:
+        """Return every observer owned by this module."""
+        ...
```
[the remark this answers is not in the document]
Comment (mhs4670go) [reply tico/experimental/quantization/ptq/wrappers/quant_module_base.py:86]: Technically speaking, observers are attached to operations rather than modules. Wrapped modules run operations, which means they has multiple observers.

`.children()` only iterates over sub-modules. `_all_observers` iterates not only such modules' observers but also own operations' observers. Therefore, only calling child's observers wouldn't be enough. 

Below is the exmaple.
```python
# QuantLlamaMLP
def _all_observers(self):
        # local first
        yield self.act_in_obs
        yield self.mul_obs
        # recurse into children that are QuantModuleBase
        for m in (self.gate_proj, self.up_proj, self.down_proj, self.act_fn):
            yield from m._all_observers()
```
